CodeJam.Main/Threading/TaskHelper.ForEachAsync.cs

```diff
@@ -0,0 +1,191 @@
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+using JetBrains.Annotations;
+
+#if NET45_OR_GREATER || TARGETS_NETSTANDARD || TARGETS_NETCOREAPP
+using TaskEx = System.Threading.Tasks.Task;
+#else
+using TaskEx = System.Threading.Tasks.TaskEx;
+#endif
+
+namespace CodeJam.Threading
+{
+	/// <summary>
+	/// Helper methods for <see cref="Task"/> and <see cref="Task{TResult}"/>.
+	/// </summary>
+	public static partial class TaskHelper
+	{
+		private const int _maxProcessorCountRefreshTicks = 30000;
+		private static volatile int _processorCount;
+		private static volatile int _lastProcessorCountRefreshTicks;
+
+		// BASEDON PlatformHelper.ProcessorCount
+		private static int ProcessorCount
+		{
+			get
+			{
+				var tickCount = Environment.TickCount;
+				if (_processorCount == 0 || tickCount - _lastProcessorCountRefreshTicks >= _maxProcessorCountRefreshTicks)
+				{
+					_processorCount = Environment.ProcessorCount;
+					_lastProcessorCountRefreshTicks = tickCount;
+				}
+				return _processorCount;
+			}
+		}
+
+		/// <summary>
+		/// Gets the maximum degree of parallelism for the scheduler.
+		/// Matches to the <see cref="Parallel.ForEach{TSource}(IEnumerable{TSource}, Action{TSource})"/> behavior.
+		/// Limits <paramref name="value"/> by <see cref="TaskScheduler.MaximumConcurrencyLevel"/> value (if non-zero positiver value).
+		/// Otherwise, uses <see cref="Environment.ProcessorCount"/> as fallback value.
+		/// </summary>
+		public static int GetMaxDegreeOfParallelism([NotNull] this TaskScheduler scheduler, int value)
+		{
+			var concurrencyLimit = scheduler.MaximumConcurrencyLevel;
+
+			if (concurrencyLimit > 0 && concurrencyLimit != int.MaxValue)
+				return value <= 0 ? concurrencyLimit : Math.Min(concurrencyLimit, value);
+
+			return value <= 0 ? ProcessorCount : value;
+		}
+
+		/// <summary>
+		/// Runs actions over source items concurrently and asynchronously.
+		/// </summary>
+		/// <typeparam name="T">Type of items to process</typeparam>
+		/// <param name="source">The source.</param>
+		/// <param name="callback">The callback.</param>
+		/// <param name="maxDegreeOfParallelism">
+		/// The maximum degree of parallelism. If zero or negative, default scheduler value is used.
+		/// See <see cref="GetMaxDegreeOfParallelism"/> documentation for more details.
+		/// </param>
+		/// <param name="cancellation">The cancellation.</param>
+		// BASEDON https://stackoverflow.com/a/25877042
+		public static async Task ForEachAsync<T>(
+			[NotNull] this IEnumerable<T> source,
+			[NotNull] Func<T, CancellationToken, Task> callback,
+			int maxDegreeOfParallelism = 0,
+			CancellationToken cancellation = default)
+		{
+			maxDegreeOfParallelism = TaskScheduler.Current.GetMaxDegreeOfParallelism(maxDegreeOfParallelism);
+
+			using (var failureCancellation = new CancellationTokenSource())
+			using (var combinedCancellation = CreateCancellation(cancellation, failureCancellation.Token))
```

Comment:
```suggestion
			using var failureCancellation = new CancellationTokenSource();
			using var combinedCancellation = CreateCancellation(cancellation, failureCancellation.Token);
```

Less nesting

Reply:
Never!:)

The using var is not the best syntax as it leaves no control over the variable scope and may result in unexpected order of dispose calls.

I've a really long debugging session one or two times due to silly mistakes in others code and I do strongly prefer clarity over compactness:)

Reply:
Control is the scope of the variable just like C++.

Reply:
Yeah. but it is a new thing for most of the c# programmers and sometimes it results in a unintended behavior such as extending transaction lifetime.

Also, i see no point in doing same thing in a multiple ways. There is a syntax for the using statement since c#1 and it has no major drawbacks (unlike the new one).

Reply:
Ok.

All my colleagues are C++ developers so they really like this syntax and they understand clearly what it does :)